source/material_model/reaction_model/grain_size_evolution.cc

```diff
@@ -0,0 +1,669 @@
+                                const std::function<double(
+                                  const double,const double,const double,const double,const double,const unsigned int)> &diffusion_viscosity,
+                                const double                                         min_eta,
+                                const double                                         max_eta,
+                                typename Interface<dim>::MaterialModelOutputs       &out) const
+      {
+        // We want to iterate over the grain size evolution here, as we solve in fact an ordinary differential equation
+        // and it is not correct to use the starting grain size (and it introduces instabilities).
+        // We assume that the strain rate is constant across all substeps for the ODE (within one advection step).
+        // Even though this assumption may not always be the most accurate (compared to assuming a constant stress),
+        // it leads to a more stable behavior because it implies that a reduction in grain size leads to less work
+        // being done by dislocation creep, so that the grain size reduction rate is lower in the next substep.
+        std::vector<std::vector<double>> &reaction_terms = out.reaction_terms;
+
+        const unsigned int n_evaluation_points = in.n_evaluation_points();
+        const unsigned int grain_size_index = this->introspection().get_indices_for_fields_of_type(CompositionalFieldDescription::grain_size)[0];
+
+        // Set up a vector that tells us which phase transition has been crossed for each point we are evaluating.
+        std::vector<int> crossed_transitions (n_evaluation_points, -1);
+
+        // Copy grain sizes into a VectorType for the ODE solver.
+        // While there also limit grain sizes to be larger than the minimum grain size.
+        using VectorType = Vector<double>;
+        VectorType grain_sizes(n_evaluation_points);
+        for (unsigned int i=0; i<n_evaluation_points; ++i)
+          grain_sizes[i] = std::max(minimum_grain_size, in.composition[i][grain_size_index]);
+
+        const double timestep = this->simulator_is_past_initialization()
+                                ?
+                                this->get_timestep()
+                                :
+                                0.0;
+
+        // If the grain sizes are not valid, we are in initialization, or the time
+        // step size is zero, we do not solve the ODE.
+        if (std::all_of(grain_sizes.begin(), grain_sizes.end(), [](double gs)
+        {
+          return gs != gs || gs < std::numeric_limits<double>::min();
+          })
+        || timestep == 0.0)
+        return;
+
+        SUNDIALS::ARKode<VectorType>::AdditionalData data;
+
+        data.initial_time = 0.0;
+        data.final_time = this->get_timestep();
+
+        // TODO: make this an input parameter.
```

Comment:
Might be worth opening an issue to fix this as a starter project and cross-reference the issue number here.